r-data-manipulation.Rmd

```diff
@@ -0,0 +1,1140 @@
+    n_licenses = n()
+  )
+```
+the `filter()` step came after the `mutate()` step. Does this matter?  
+
+* Swap the order in the code and see if you get the same results.
+
+* Write out how you might describe the steps. Is it obvious you can swap the filter and mutate step and get the same results?
+
+* Which steps can't you swap the order of? Why?
+
+* *Despite giving the same results, some orderings of the data manipulation steps will take longer to compute. Can you guess why?*
+
+### Exercise: the five boroughs
+
+The column `neighborhood_tabulation_area` is a code for the "Neighborhood Tabulation Areas", and has been geo-coded from the licensee's address (as opposed to self reported).  The first two characters correspond to the Borough.
+
+This code creates a new variable called `borough_code` that contains just these two characters:
+```{r}
+dog_licenses %>% 
+  mutate(borough_code = stringr::str_sub(neighborhood_tabulation_area, 1, 2)) 
+```
+
+Which borough has the longest average licence duration?
+
+## How can I work with two or more datasets?
+
+So far all the manipulation verbs you have seen work on a single table of data.  But what if your questions involve data spread across more than one dataset?  As an example, consider the question: which neighborhood has the highest proportion of dog ownership in 2016?  You can get a handle on how many dog are licensed during 2016 for each neighborhood using the manipulations you've seen so far:
+```{r}
+dogs_by_nta <- dog_licenses %>% 
+  filter(
+    license_issued_date >= as.Date("2016-01-01"), 
+    license_issued_date <= as.Date("2016-12-31")
+    ) %>% 
+  group_by(neighborhood_tabulation_area) %>% 
+  summarise(n_licences = n()) %>% 
+  arrange(desc(n_licences))
+dogs_by_nta
+```
+
+Looks like MN12 has the most licenses, but what neighborhood is this?  And are there many licenses because a high proportion of people own dogs, or just that there are lots of people in this neighborhood?  To answer both of these questions you would need to combine this data with additional data, say that contained in `nyc-nta-pop`
+```{r, eval = FALSE}
+nta_pop <- read_csv("nyc-nta-population.csv")
+```
+```{r, echo = FALSE}
+nta_pop <- read_csv(here("data", "nyc-nta-population.csv"))
+```
+
+```{r}
+nta_pop
+```
+
+This data includes populations for each NTA and a full name for the NTA.  If you look at the neighborhoods with the largest populations:
+```{r}
+nta_pop %>% 
+  arrange(desc(population)) %>% 
+  select(nta_code, nta_name, year, population, year)
+```
+You'll see that MN12, the Upper West Side, in fact has the largest population of all the neighborhoods (at least in 2000 and 2010), so perhaps it isn't surprising it also has the most dog licenses issued in 2016.
+
+Joins are a way to combine two datasets together by using a column the datasets have in common - like the columns that contain the NTA codes.  Let's work with some smaller datasets to see how this works.  
```

Comment:
I would be inclined to introduce append/concat functionality (I think `bind_row`/`bind_col` in dply?) before joins since these are simply adding rows or columns which I think is simpler to understand than adding based on the values in a common column.